sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala

```diff
@@ -59,9 +59,14 @@ class QuantileSummaries(
    * @param x the new observation to insert into the summary
    */
   def insert(x: Double): QuantileSummaries = {
-    headSampled.append(x)
+    headSampled += x
     if (headSampled.size >= defaultHeadSize) {
-      this.withHeadBufferInserted
+      val result = this.withHeadBufferInserted
+      if (result.sampled.length >= compressThreshold) {
+        result.compress()
```

Comment:
@clockfly oh, hm, this code just landed recently? it seems to call `compress()` itself rather than leave it to `QuantileSummaries`, in which case I'm not clear why there's a `compressThreshold` in `QuantileSummaries` It seems like the new class is trying to manage it. What's the right way to rationalize this -- are you saying `QuantileSummaries` shouldn't manage compression at all? that's fine too (in which case this can just turn into a very small optimization change).